### PR TITLE
EASY-1972: Implementeer easy-licenses als Maven artifact

### PR DESCRIPTION
**Pull request by vesaakerman**
fixes EASY-1972

#### When applied it will
* Implement list of Easy licenses as a Maven articraft

#### related pull requests on github
repo                       | PR
-------------------------- | -----------------
easy-stage-dataset | [PR#1973](https://github.com/DANS-KNAW/easy-stage-dataset/pull/143) 
easy-ingest-flow | [PR#1974](https://github.com/DANS-KNAW/easy-ingest-flow/pull/120) 
easy-split-multi-deposit | [PR#1976](https://github.com/DANS-KNAW/easy-split-multi-deposit/pull/139) 
easy-validate-dans-bag| [PR#1977](https://github.com/DANS-KNAW/easy-validate-dans-bag/pull/65)

### CODE REVIEW

File: pom.xml [hunk 1]
```diff
@@ -0,0 +1,172 @@
...
+        <version>4.0.0</version>
+    </parent>
+
+    <groupId>nl.knaw.dans.easy</groupId>
+    <artifactId>easy-licenses</artifactId>
+    <version>1.0.0-SNAPSHOT</version>
+
+    <name>EASY licenses</name>
+    <description>Collection of EASY licenses</description>
+    <inceptionYear>2012</inceptionYear>
```
Comment (vesaakerman) [inline pom.xml]: If I change the year to something greater that 2012, I get an error message:
`Some files do not have the expected license header`
????

Comment (rvanheest) [reply pom.xml]: It's the license headers... Run

    mvn license:format

after you changed the year

File: .gitignore [hunk 2]
```diff
@@ -0,0 +1,29 @@
...
+/.vagrant/
+/.yum-repo/
+/.dans.knaw.nl-yum-repo
+
+# Eclipse
+/.project
+/.settings
+/.classpath
+/.cache
+
```
Comment (rvanheest) [inline .gitignore:4]:
```suggestion
```

File: pom.xml [hunk 3]
```diff
@@ -0,0 +1,172 @@
...
+
+    <scm>
+        <developerConnection>scm:git:https://github.com/DANS-KNAW/${project.artifactId}</developerConnection>
+        <tag>HEAD</tag>
+    </scm>
+
+    <dependencies>
```
Comment (rvanheest) [inline pom.xml:41]: you don't have tests, so these dependencies are not necessary to be added

File: pom.xml [hunk 4]
```diff
@@ -0,0 +1,172 @@
...
+    See the License for the specific language governing permissions and
+    limitations under the License.
+
+-->
+<project xmlns="http://maven.apache.org/POM/4.0.0" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xsi:schemaLocation="http://maven.apache.org/POM/4.0.0 http://maven.apache.org/xsd/maven-4.0.0.xsd">
+    <modelVersion>4.0.0</modelVersion>
+
+    <parent>
+        <groupId>nl.knaw.dans.shared</groupId>
+        <artifactId>dans-scala-project</artifactId>
```
Comment (rvanheest) [inline pom.xml:24]: @janvanmansum could you check whether this is the correct parent project for this?

File: pom.xml [hunk 5]
```diff
@@ -0,0 +1,172 @@
...
+            <resource>
+                <directory>src/main/assembly/dist</directory>
+                <targetPath>${project.build.directory}/version</targetPath>
+                <includes>
+                    <include>version</include>
+                </includes>
+                <filtering>true</filtering>
+            </resource>
+        </resources>
+        <plugins>
```
Comment (rvanheest) [inline pom.xml:93]: No code here, so `scala-maven-plugin` and `scalatest-maven-plugin` are probably not required

File: pom.xml [hunk 6]
```diff
@@ -0,0 +1,172 @@
...
+                <groupId>org.scalatest</groupId>
+                <artifactId>scalatest-maven-plugin</artifactId>
+            </plugin>
+            <plugin>
+                <artifactId>maven-assembly-plugin</artifactId>
+            </plugin>
+            <plugin>
+                <groupId>com.mycila</groupId>
+                <artifactId>license-maven-plugin</artifactId>
+                <configuration combine.children="append">
```
Comment (rvanheest) [inline pom.xml]: you copied lots of stuff, but this is all unnecessary. There are no `*.xsd` files, neither is there a directory `src/test/scala/**`.

File: pom.xml [hunk 7]
```diff
@@ -0,0 +1,172 @@
...
+                        <artifactId>maven-antrun-plugin</artifactId>
+                    </plugin>
+                    <plugin>
+                        <groupId>org.codehaus.mojo</groupId>
+                        <artifactId>rpm-maven-plugin</artifactId>
+                        <configuration>
+                            <group>Applications/Archiving</group>
+                            <mappings combine.self="override">
+                                <mapping>
+                                    <directory>/var/www/html/licenses</directory>
```
Comment (rvanheest) [inline pom.xml]: I don't think we want to install or expose these licenses via this path.
In fact, I'm most certain this `pom.xml` file doesn't even need a `<profiles>`.

File: src/main/assembly/dist/version [hunk 8]
```diff
@@ -0,0 +1 @@
+${project.version}
```
Comment (rvanheest) [inline src/main/assembly/dist/version:1]: @janvanmansum is this file necessary if we don't create an RPM of this?

**Review by rvanheest — CHANGES_REQUESTED** [review]
1. I think you copy-pasted a lot of stuff in `pom.xml` that is not necessary.
2. Please read the issue once more... I'm missing a number of files that we specified there.

File: pom.xml (hunk 4 above)
Comment (janvanmansum) [reply pom.xml:24]: We should probably use the parent highest in the hierarchy that is applicable, which would be dans-mvn-plugin-defaults.

File: src/test/scala/nl/knaw/dans/easy/licenses/LicensesJsonSpec.scala [hunk 9]
```diff
@@ -0,0 +1,42 @@
...
+  val files = new File(LICENSES_DIR).listFiles.filter(_.isFile).map(_.getName).filter(n => !n.endsWith("properties") && !n.endsWith("json")).toList
+  val fileNames = files.map(f => f.substring(0, f.lastIndexOf(".")))
+  val json = parse(new File(LICENSES_DIR, "licenses.json"))
+  val viewNames = for {
+    JObject(file) <- json
+    JField("title", viewName) <- file
+  } yield viewName.values
+
+  "all the files in licenses.json" should "be present in the licenses directory" in {
+    viewNames.foreach(fileNames should contain(_))
```
Comment (rvanheest) [inline src/test/scala/nl/knaw/dans/easy/licenses/LicensesJsonSpec.scala]: In your tests (like this one) you use `foreach`. While this works perfectly, it will fail the test on the first instance in `viewNames` it finds for which the expectation fails. If you look at

https://github.com/DANS-KNAW/easy-deposit-api/blob/80861017ade01760382fe7a643242682b26824d4/src/test/scala/nl.knaw.dans.easy.deposit/servlets/UploadSpec.scala#L58-L60

you'll see `forEvery` being used. This method collects all failing expectations and reports all of them at once, giving you a better overview of what errors occur.
For this method to work, you need to add `with Inspectors` to the class definition.

File: pom.xml [hunk 10]
```diff
@@ -0,0 +1,112 @@
...
+
+    <scm>
+        <developerConnection>scm:git:https://github.com/DANS-KNAW/${project.artifactId}</developerConnection>
+        <tag>HEAD</tag>
+    </scm>
+
+    <dependencies>
+        <dependency>
+            <groupId>commons-configuration</groupId>
+            <artifactId>commons-configuration</artifactId>
```
Comment (rvanheest) [inline pom.xml:44]: please add test scope here as well... otherwise you get this dependency in the final jar.

File: pom.xml [hunk 11]
```diff
@@ -0,0 +1,112 @@
...
+            <scope>test</scope>
+        </dependency>
+        <dependency>
+            <groupId>nl.knaw.dans.lib</groupId>
+            <artifactId>dans-scala-lib_2.12</artifactId>
+            <scope>test</scope>
+        </dependency>
+        <dependency>
+            <groupId>org.json4s</groupId>
+            <artifactId>json4s-native_2.12</artifactId>
```
Comment (rvanheest) [inline pom.xml:64]: same here...

File: pom.xml (hunk 6 above)
Comment (rvanheest) [reply pom.xml]: I think this configuration is not necessary. The extra `<includes>` you add here should already be included in the default configuration in the parent projects

File: src/test/scala/nl/knaw/dans/easy/licenses/LicensesPropertiesSpec.scala [hunk 12]
```diff
@@ -0,0 +1,39 @@
...
+
+import org.apache.commons.configuration.PropertiesConfiguration
+import org.scalatest.{ FlatSpec, Matchers }
+
+import scala.collection.JavaConverters._
+
+class LicensesPropertiesSpec extends FlatSpec with Matchers {
+
+  val LICENSES_DIR = "src/main/assembly/dist/licenses"
+  val files = new File(LICENSES_DIR).listFiles.filter(_.isFile).map(_.getName).filter(n => !n.endsWith("properties") && !n.endsWith("json")).toList
```
Comment (rvanheest) [inline src/test/scala/nl/knaw/dans/easy/licenses/LicensesPropertiesSpec.scala]: Probably you can use `filterNot` here instead of `filter`, which will give you an easier to read predicate.

File: pom.xml (hunk 1 above)
Comment (vesaakerman) [reply pom.xml]: ok, thanks!

File: pom.xml (hunk 5 above)
Comment (vesaakerman) [reply pom.xml:93]: after all, tests have been added

File: pom.xml (hunk 6 above)
Comment (vesaakerman) [reply pom.xml]: you are completely right

File: pom.xml [hunk 13]
```diff
@@ -0,0 +1,114 @@
...
+    limitations under the License.
+
+-->
+<project xmlns="http://maven.apache.org/POM/4.0.0" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xsi:schemaLocation="http://maven.apache.org/POM/4.0.0 http://maven.apache.org/xsd/maven-4.0.0.xsd">
+    <modelVersion>4.0.0</modelVersion>
+
+    <parent>
+        <groupId>nl.knaw.dans.shared</groupId>
+        <artifactId>dans-scala-project</artifactId>
+        <version>4.0.0</version>
```
Comment (rvanheest) [inline pom.xml]:
```suggestion
        <version>5.0.0</version>
```

File: pom.xml [hunk 14]
```diff
@@ -0,0 +1,114 @@
...
+            <scope>test</scope>
+        </dependency>
+        <dependency>
+            <groupId>org.json4s</groupId>
+            <artifactId>json4s-native_2.12</artifactId>
+            <scope>test</scope>
+        </dependency>
+    </dependencies>
+
+    <repositories>
```
Comment (rvanheest) [inline pom.xml:69]: these have changed over the last couple of days. see any other project for the `<repositories>` tag; the `<pluginRepositories>` tag is fully gone.

File: pom.xml [hunk 15]
```diff
@@ -0,0 +1,107 @@
...
+            </releases>
+            <snapshots>
+                <enabled>true</enabled>
+            </snapshots>
+            <url>https://maven.dans.knaw.nl/snapshots/</url>
+        </repository>
+    </repositories>
+
+    <build>
+        <plugins>
```
Comment (janvanmansum) [inline pom.xml:93]:
```suggestion
        <plugins>
            <plugin>
                <artifactId>maven-jar-plugin</artifactId>
                <executions>
                    <execution>
                        <id>default-jar</id>
                        <phase>none</phase>
                    </execution>
                </executions>
            </plugin>
```

In order to avoid the warning about an empty jar being created.

File: src/test/scala/nl/knaw/dans/easy/licenses/LicensesJsonSpec.scala [hunk 16]
```diff
@@ -0,0 +1,43 @@
...
+
+import java.io.File
+
+import org.json4s._
+import org.json4s.native.JsonMethods._
+import org.scalatest.{ FlatSpec, Inspectors, Matchers }
+
+class LicensesJsonSpec extends FlatSpec with Matchers with Inspectors {
+
+  val LICENSES_DIR = "src/main/assembly/dist/licenses"
```
Comment (janvanmansum) [inline src/test/scala/nl/knaw/dans/easy/licenses/LicensesJsonSpec.scala]: Mind the warnings in IntelliJ. Make these vals private.

File: .gitignore (hunk 2 above)
Comment (vesaakerman) [reply .gitignore:4]: done

Comment (rvanheest) [reply .gitignore:4]: It didn't change yet? Maybe you forgot to push your commit?
